src/main/scala/org/broadinstitute/hail/asm4s/FunctionBuilder.scala

```diff
@@ -88,110 +88,402 @@ abstract class FunctionBuilder[R](parameterTypeInfo: Array[TypeInfo[_]], returnT
+    }
+  }
+}
 
-  def arg2 = getArg[A2](1)
+trait IToI { def apply(b: Int): Int }
+class FunctionIToIBuilder(implicit act: ClassTag[Int], ati: TypeInfo[Int],
+  rti: TypeInfo[Int]) extends FunctionBuilder[Int](Array[TypeInfo[_]](ati), rti) {
+
+  cn.interfaces.asInstanceOf[java.util.List[String]].add("org/broadinstitute/hail/asm4s/IToI")
+
+  def arg1 = getArg[Int](1)
+
+  def result(c: Code[Int]): (Int) => Int = {
+    val bytes = classAsBytes(c)
+    val localName = name.replaceAll("/",".")
+
+    new Function1[Int, Int] with java.io.Serializable {
+      @transient @volatile private var f: IToI = null
+      def apply(a: Int): Int = {
+        if (f == null) {
+          this.synchronized {
+            if (f == null) {
+              f = loadClass(localName, bytes).newInstance().asInstanceOf[IToI]
+            }
+          }
+        }
+
+        f(a)
+      }
+    }
+  }
+}
+
+
+trait AToI[A] { def apply(b: A): Int }
+class FunctionAToIBuilder[A](implicit act: ClassTag[A], ati: TypeInfo[A],
+  rti: TypeInfo[Int]) extends FunctionBuilder[Int](Array[TypeInfo[_]](implicitly[TypeInfo[AnyRef]]), rti) {
+
+  cn.interfaces.asInstanceOf[java.util.List[String]].add("org/broadinstitute/hail/asm4s/AToI")
+
+  def arg1 = Code.checkcast[A](getArg[AnyRef](1))
+
+  def result(c: Code[Int]): (A) => Int = {
+    val bytes = classAsBytes(c)
+    val localName = name.replaceAll("/",".")
+
+    new Function1[A, Int] with java.io.Serializable {
+      @transient @volatile private var f: AToI[A] = null
+      def apply(a: A): Int = {
+        if (f == null) {
+          this.synchronized {
+            if (f == null) {
+              f = loadClass(localName, bytes).newInstance().asInstanceOf[AToI[A]]
+            }
+          }
+        }
+
+        f(a)
+      }
+    }
+  }
+}
+
+class Function2Builder[A1 >: Null, A2 >: Null, R >: Null]
+  (implicit a1ct: ClassTag[A1], a1ti: TypeInfo[A1], a2ct: ClassTag[A2], a2ti: TypeInfo[A2], rti: TypeInfo[R])
+    extends FunctionBuilder[R](Array[TypeInfo[_]](implicitly[TypeInfo[AnyRef]], implicitly[TypeInfo[AnyRef]]), rti) {
+
+  cn.interfaces.asInstanceOf[java.util.List[String]].add("scala/Function2")
+
+  def arg1 = Code.checkcast[A1](getArg[AnyRef](1))
+
+  def arg2 = Code.checkcast[A2](getArg[AnyRef](2))
 
   def result(c: Code[R]): (A1, A2) => R = {
-    val clazz = resultClass(c)
-    val m = clazz.getMethod("f", a1ct.runtimeClass, a2ct.runtimeClass)
-    (a1: A1, a2: A2) => m.invoke(null, a1.asInstanceOf[AnyRef], a2.asInstanceOf[AnyRef]).asInstanceOf[R]
+    val bytes = classAsBytes(c)
+    val localName = name.replaceAll("/",".")
+
+    new Function2[A1, A2, R] with java.io.Serializable {
+      @transient @volatile private var f: (Any, Any) => Any = null
+      def apply(a1: A1, a2: A2): R = {
+        if (f == null) {
+          this.synchronized {
+            if (f == null) {
+              f = loadClass(localName, bytes).newInstance().asInstanceOf[(Any, Any) => Any]
+            }
+          }
+        }
```

Comment:
Double locking, AFAIK, is correct with `@volatile`. I'm not sure it even matters though. Would be nice to benchmark double-checked locking against regular old, simple single locking.